codeU3.py

```diff
@@ -0,0 +1,222 @@
+        self.assertTrue(self.d.isPrefix('DOGS'))
+
+    def test_normal_input(self):
+        self.assertTrue(self.d.isWord('cats'))
+        self.assertTrue(self.d.isPrefix('car'))
+
+        self.assertFalse(self.d.isWord('cat'))
+        self.assertFalse(self.d.isPrefix('cs'))
+
+    def test_nonstring_input(self):
+        self.assertRaises(self.d.isWord(['cat', 'cats']))
+        self.assertRaises(self.d.isPrefix(['cat', 'cats']))
+
+class CharGrid(object):
+
+    '''Given (n by m) grid of letters and a dictionary
+    Return all possible words from dictionary that can be formed by traversing
+    the grid cells without visiting the same cell twice
+    '''
+
+    def __init__(self, char_matrix, word_list):
+        
+        '''Given char_matrix as 2D numpy array and a list of words
+        Checks whether we have a proper matrix and character values in it
+        and saves them along with matrix shape as attributes
+        '''
+        
+        char_array = char_matrix.flatten() # flattening for checking matrix shape correctness
+        for c in char_array:
+            if isinstance(c, list):
+                assert TypeError, 'Please provide matrix type of data'
+                return
+            if not c.isalpha() or len(c) > 1:
+                assert TypeError, 'Please provide grid values with only char entries'
+                return
+        self.grid = char_matrix
+        self.dict = Dictionary(word_list)
+        (self.m, self.n) = np.shape(char_matrix)
+
+    def is_route_prefix(self, route):
+
+        '''Given a list of tuple coordinates (x,y) encoding route
+        Returns:
+            True: if route gives string that is a prefix in our dictionary
+            False: otherwise
+        '''
+
+        if not route:
+            return False
+        chars = [self.grid[x, y] for (x, y) in route]
+        prefix = ''.join(chars)
+        if self.dict.isPrefix(prefix):
+            return True
+        return False
+
+    def route_word(self, route):
+
+        '''Given a list of tuple coordinates (x,y) encoding route
+        Returns:
+            empty string, if corresponding string is not in the dictionary
+            word, otherwise
+        '''
+
+        if not route:
+            return ""
+        chars = [self.grid[x, y] for (x, y) in route]
+        word = ''.join(chars)
+        if self.dict.isWord(word):
+            return word
+        return ""
+
+    def route_extend(self, route):
+
+        '''Given a route as list of (x,y) indices
+        Returns:
+            all possible routes where 1 step made in one of 8 directions
+            each route is a list of tuple coordinates 
+        '''
+
+        last_x = route[-1][0] #last route coordinate x
+        last_y = route[-1][1] #last route coordinate y
+        cells_set = set(route)
+        route_extended = []
+
+        if last_x + 1 < self.m and (last_x + 1, last_y) not in cells_set: #ensure we don't visit the same cell twice

```

Comment:
It could be the right choice but most people would also consider some kind of loop for this. Do you have opinions why it's best to fully write all this conditions out?